Add CKE-specific chai assertions

## 📝 Provide a description of the improvement

It would make writing tests easier, assertions more readable, and fail messages more meaningful.

Take https://github.com/ckeditor/ckeditor5/pull/7356/commits/8888fcebde4516d60432a17706750b7e67fa28fe#diff-74026e14057514eeea5cabde2b2f809eR16-R31 as an example

Thanks to that you can write
```js
expect( model.document.selection ).to.have.an.attribute( 'linkHref' );
// AssertionError: expected { Object (_selection) } to have 'linkHref' attribute
```
instead of 
```js
expect( model.document.selection.hasAttribute( 'linkHref' ) ).to.be.true;
//  AssertionError: expected false to be true
```

So I have two questions:
- Does anybody have anything against such an approach?
- Where to store such assertions? - `-dev-tests`?


## 📃 Other details

* Browser: n/a
* OS: n/a
* CKEditor version: [master@b80c677949ea4](https://github.com/ckeditor/ckeditor5/tree/b80c677949ea4736810a2e17f0cc19b2f3cf34a7)
* Installed CKEditor plugins: n/a

---

If you'd like to see this improvement implemented, add a 👍 reaction to this post.

About Questions:

> *   Does anybody have anything against such an approach?

Nope :) The solution is OK for me. I think that Assertion error messages could be smarter. Something like this (or anything that is fast to write)

```javascript
expect( model.document.selection ).to.have.an.attribute( 'linkHref' );
// AssertionError: expected model:documentSelection to have 'linkHref' attribute

expect( paragraph ).to.have.an.attribute( 'linkHref' );
// AssertionError: expected model:element (paragraph) to have 'linkHref' attribute

// fall-back:
expect( { foo: 'bar' } ).to.have.an.attribute( 'linkHref' );
// AssertionError: expected { Object( foo ) } to have 'linkHref' attribute
```

> *   Where to store such assertions? - `-dev-tests`?

ATM: `ckeditor5-core/tests/_utils/`.

Desired: `ckeditor5-test-utils` (#6313, #6657).

> I think that Assertion error messages could be smarter. [...]

What you have shown here, is not a part of the assertion, but Chai's way to display object's and their names https://github.com/chaijs/chai/blob/058ddadb8422238b418d0c3e8f92e4f757289abd/lib/chai/utils/objDisplay.js#L27-L50


The way `sinon-chai` works it around, their check if the object in question is a spy, and puts the name defined in the spy into the message. Nice trick, but still requires a name to be defined per-object.


BTW, I wonder why Chai is not using the `obj.constructor.name` here, so:
```js
expect( model.document.selection ).to.have.an.attribute( 'linkHref' );
// AssertionError: expected { DocumentSelection(_selection) } to have 'linkHref' attribute
```
Would that be any batter for you? If so, I'd make a PR to chai.

--- a/packages/ckeditor5-core/tests/_utils-tests/assertions/attribute.js
+++ b/packages/ckeditor5-core/tests/_utils-tests/assertions/attribute.js
@@ -1,0 +1,79 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* global chai */
+
+import '../../_utils/assertions/attribute';
+
+describe( 'attribute chai assertion', () => {
+	it( 'should be added to chai assertions', () => {
+		const assertion = new chai.Assertion();
+
+		expect( assertion ).to.have.property( 'attribute' );
+		expect( assertion.attribute ).to.be.instanceof( Function );
+	} );
+
+	it( 'should assert the target has a \'hasAttribute\' method', () => {
+		expect( { hasAttribute: () => true } ).to.have.attribute( 'foo' );
+
+		expect( function() {
+			expect( {} ).not.to.have.attribute( 'bar' );
+		} ).to.throw( 'expected {} to respond to \'hasAttribute\'' );
+
+		expect( function() {
+			expect( {} ).to.have.attribute( 'bar' );
+		} ).to.throw( 'expected {} to respond to \'hasAttribute\'' );
+	} );
+
+	it( 'should assert the \'target.hasAttribute\' returns \'true\' for the given type', () => {
+		expect( { hasAttribute: () => true } ).to.have.attribute( 'foo' );
+
+		expect( function() {
+			expect( { hasAttribute: () => false } ).to.have.attribute( 'bar' );
+		} ).to.throw( 'expected { Object (hasAttribute) } to have attribute \'bar\'' );
+	} );
+
+	it( 'negated, should assert the \'target.hasAttribute\' returns \'false\' for the given type', () => {
+		expect( { hasAttribute: () => false } ).not.to.have.attribute( 'foo' );
+
+		expect( function() {
+			expect( { hasAttribute: () => true } ).not.to.have.attribute( 'bar' );
+		} ).to.throw( 'expected { Object (hasAttribute) } to not have attribute \'bar\'' );
+	} );
+
+	it( 'should assert the \'target.getAttribute\' returns the given value for the given type', () => {
+		expect( {
+			hasAttribute: () => true,
+			getAttribute: () => 'bar'
+		} ).to.have.attribute( 'foo', 'bar' );
+
+		expect( function() {
+			expect( {
+				hasAttribute: () => true,
+				getAttribute: () => 'bar'
+			} ).to.have.attribute( 'foo', 'baz' );
+		} ).to.throw( 'expected { Object (hasAttribute, getAttribute) } to have attribute \'foo\' of \'bar\', but got \'baz\'' );
+	} );
+
+	it( 'negated, should assert for the given type the \'target.getAttribute\' returns a value different than the given one', () => {
+		expect( {
+			hasAttribute: () => true,
+			getAttribute: () => 'bar'
+		} ).to.not.have.attribute( 'foo', 'baz' );
+
+		expect( function() {
+			expect( {
+				hasAttribute: () => true,
+				getAttribute: () => 'baz'
+			} ).to.not.have.attribute( 'foo', 'baz' );
+		} ).to.throw( 'expected { Object (hasAttribute, getAttribute) } to not have attribute \'foo\' of \'baz\'' );
+	} );
+
+	it( 'should prefix failure message with the given one', () => {
+		expect( function() {
+			expect( {} ).to.have.attribute( 'foo', 'baz', 'Illegal salmon' );
+		} ).to.throw( /^Illegal salmon: / );
+	} );
+} );

--- a/packages/ckeditor5-core/tests/_utils/assertions/attribute.js
+++ b/packages/ckeditor5-core/tests/_utils/assertions/attribute.js
@@ -1,0 +1,61 @@
+/**
+ * @license Copyright (c) 2003-2020, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/* global chai */
+
+/**
+ * Asserts that the target has an attribute with the given key name.
+ * See {@link module:engine/model/documentselection~DocumentSelection#hasAttribute hasAttribute}.
+ *
+ *		expect( selection ).to.have.attribute( 'linkHref' );
+ *
+ * When `value` is provided, .attribute also asserts that the attribute's value is equal to the given `value`.
+ * See {@link module:engine/model/documentselection~DocumentSelection#getAttribute getAttribute}.
+ *
+ *		expect( selection ).to.have.attribute( 'linkHref', 'example.com' );
+ *
+ * Negations works as well.
+ *
+ * @param {String} key Key of attribute to assert.
+ * @param {String} [value] Attribute value to assert.
+ * @param {String} [message] Additional message.
+ */
+chai.Assertion.addMethod( 'attribute', function attributeAssertion( key, value, message ) {
+	if ( message ) {
+		chai.util.flag( this, 'message', message );
+	}
+
+	const obj = this._obj;
+
+	if ( arguments.length === 1 ) {
+		// Check if it has the method at all.
+		new chai.Assertion( obj ).to.respondTo( 'hasAttribute' );
+
+		// Check if it has the attribute.
+		const hasAttribute = obj.hasAttribute( key );
+		this.assert(
+			hasAttribute === true,
+			`expected #{this} to have attribute '${ key }'`,
+			`expected #{this} to not have attribute '${ key }'`,
+			!chai.util.flag( this, 'negate' ),
+			hasAttribute
+		);
+	}
+
+	// If a value was given.
+	if ( arguments.length >= 2 ) {
+		// Check if it has the method at all.
+		new chai.Assertion( obj ).to.respondTo( 'getAttribute', message );
+
+		const attributeValue = obj.getAttribute( key );
+		this.assert(
+			attributeValue === value,
+			`expected #{this} to have attribute '${ key }' of #{exp}, but got #{act}`,
+			`expected #{this} to not have attribute '${ key }' of #{exp}`,
+			attributeValue,
+			value
+		);
+	}
+} );

